src/x11/ewmh.cpp

```diff
@@ -84,6 +84,14 @@ namespace ewmh_util {
     return desktops;
   }
 
+  unsigned int get_number_of_windows(unsigned int desktop) {
```

Comment:
I think getting the number of windows per workspace is possible in a single iteration over the client list instead of doing it once per workspace.

We already iterate over the client list in `rebuild_clientlist`. I would suggest using that loop to also build a mapping from workspace to number of windows, similarly to `m_clients`. When replacing the tokens you would just need to read from the map.